executor/adapter.go

```diff
@@ -1263,7 +1264,10 @@ func (a *ExecStmt) observeStmtBeginForTopSQL() {
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		costTime := time.Since(vars.StartTime) + vars.DurationParse
+		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest, func(item *stmtstats.StatementStatsItem) {
+			item.SumExecNanoDuration += uint64(costTime.Nanoseconds())
+		})
```

Comment:
I think you should write your logic inside `OnExecutionFinished`.

Reply:
Great, done.